File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/form/Button.java
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2017 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
package com.adobe.cq.wcm.core.components.models.form;

import org.osgi.annotation.versioning.ConsumerType;

/**
 * Defines the {@code Button} Sling Model used for the {@code /apps/core/wcm/components/button} component.
 *
 * @since com.adobe.cq.wcm.core.components.models.form 13.0.0
 */
@ConsumerType
public interface Button extends Field {

    /**
     * Defines button type.
     *
     * @since com.adobe.cq.wcm.core.components.models.form 13.0.0
     */
    enum Type {
        /**
         * Button type used to submit forms.
         *
         * @since com.adobe.cq.wcm.core.components.models.form 13.0.0
         */
        SUBMIT("submit"),

        /**
         * Normal button.
         *
         * @since com.adobe.cq.wcm.core.components.models.form 13.0.0
         */
        BUTTON("button");

        private String value;

        Type(String value) {
            this.value = value;
        }

        /**
         * Given a {@link String} <code>value</code>, this method returns the enum's value that corresponds to the provided string
         * representation. If no representation is found, {@link #SUBMIT} will be returned.
         *
         * @param value the string representation for which an enum value should be returned
         * @return the corresponding enum value, if one was found, or {@link #SUBMIT}
         * @since com.adobe.cq.wcm.core.components.models.form 13.0.0
         */
        public static Type fromString(String value) {
            for (Type type : values()) {
                if (type.value.equals(value)) {
                    return type;
                }
            }
            return SUBMIT;
        }
    }

    /**
     * Returns this button's type.
     *
     * @return the type of the button; possible values: 'button', 'submit'
     * @since com.adobe.cq.wcm.core.components.models.form 13.0.0; marked <code>default</code> in 14.1.0
     */
    default Type getType() {
        return null;
    }

}

File: content/src/content/jcr_root/apps/core/wcm/components/button/v2/button/clientlibs/editor/js/button.js
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 ~ Copyright 2022 Adobe
 ~
 ~ Licensed under the Apache License, Version 2.0 (the "License");
 ~ you may not use this file except in compliance with the License.
 ~ You may obtain a copy of the License at
 ~
 ~     http://www.apache.org/licenses/LICENSE-2.0
 ~
 ~ Unless required by applicable law or agreed to in writing, software
 ~ distributed under the License is distributed on an "AS IS" BASIS,
 ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ~ See the License for the specific language governing permissions and
 ~ limitations under the License.
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
(function($) {
    "use strict";

    var selectors = {
        dialogContent: ".cmp-button__editor",
        linkField: 'input[name="./link"]',
        linkURLField: 'foundation-autocomplete[name="./linkURL"]'
    };
    var $linkField;
    var $linkURLField;
    var link;
    var linkURL;

    $(document).on("dialog-loaded", function(event) {
        var $dialog = event.dialog;
        var $dialogContent = $dialog.find(selectors.dialogContent);
        var dialogContent  = $dialogContent.length > 0 ? $dialogContent[0] : undefined;
        if (dialogContent) {
            $linkField = $dialogContent.find(selectors.linkField);
            $linkURLField = $dialogContent.find(selectors.linkURLField);
            displayLinkProperty();
        }
    });

    /**
     * Displays the linkURL or link property if available and removes the link property when the dialog is submitted.
     */
    function displayLinkProperty() {
        if ($linkField && $linkURLField && $linkURLField.adaptTo("foundation-field")) {
            link = $linkField.val();
            linkURL = $linkURLField.adaptTo("foundation-field").getValue();
            // if the 'link' property is set and the 'linkURL' is not set: display its value in the 'linkURL' field
            if (!linkURL && link) {
                $linkURLField.adaptTo("foundation-field").setValue(link);
            }
            // remove the 'link' property in the JCR repository
            link = $linkField.val("");
        }
    }

})(jQuery);
